android/java/org/chromium/chrome/browser/crypto_wallet/activities/AddAssetActivity.java

```diff
@@ -0,0 +1,258 @@
+        setContentView(R.layout.activity_add_asset);
+
+        mToolbar = findViewById(R.id.toolbar);
+        mToolbar.setTitle(mNftsOnly ? R.string.wallet_add_nft : R.string.wallet_add_custom_asset);
+        setSupportActionBar(mToolbar);
+        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
+
+        mNetworkSpinner = findViewById(R.id.network_spinner);
+        mNetworkAdapter = new NetworkSpinnerAdapter(this, Collections.emptyList());
+        mNetworkSpinner.setAdapter(mNetworkAdapter);
+        mTokenIdSection = findViewById(R.id.token_id_section);
+        mTokenDecimalSection = findViewById(R.id.token_decimal_section);
+        mTvDecimalTitle = findViewById(R.id.token_decimals_title);
+        mTokenDecimalsEdit = findViewById(R.id.token_decimals);
+        mTvAddressTitle = findViewById(R.id.token_contract_address_title);
+
+        mAdd = findViewById(R.id.add);
+        mTokenNameEdit = findViewById(R.id.token_name);
+        mTokenContractAddressEdit = findViewById(R.id.token_contract_address);
+        mTokenSymbolEdit = findViewById(R.id.token_symbol);
+        mTokenIdEdit = findViewById(R.id.token_id);
+
+        onInitialLayoutInflationComplete();
+    }
+
+    @Override
+    public void onStartWithNative() {
+        super.onStartWithNative();
+
+        try {
+            BraveActivity activity = BraveActivity.getBraveActivity();
+            mWalletModel = activity.getWalletModel();
+            mNetworkModel = mWalletModel.getNetworkModel();
+        } catch (BraveActivity.BraveActivityNotFoundException e) {
+            Log.e(TAG, "Error onStartWithNative.", e);
+        }
+
+        LiveDataUtil.observeOnce(mNetworkModel.mCryptoNetworks,
+                networkInfoList -> { mNetworkAdapter.setNetworks(networkInfoList); });
+
+        mNetworkSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
+            @Override
+            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
+                mNetworkInfo = mNetworkAdapter.getNetwork(position);
+
+                mTokenDecimalSection.setVisibility(mNftsOnly ? View.GONE : View.VISIBLE);
+                if (mNetworkInfo.coin == CoinType.ETH || mNetworkInfo.coin == CoinType.FIL) {
+                    mTokenIdSection.setVisibility(mNftsOnly ? View.VISIBLE : View.GONE);
+                    mTvAddressTitle.setText(R.string.wallet_add_custom_asset_token_address);
+                } else if (mNetworkInfo.coin == CoinType.SOL) {
+                    mTokenIdSection.setVisibility(View.GONE);
+                    mTvAddressTitle.setText(R.string.wallet_add_custom_asset_mint_address);
+                }
+            }
+            @Override
+            public void onNothingSelected(AdapterView<?> parent) { /* No-op. */
+            }
+        });
+
+        mAdd.setOnClickListener(v -> {
+            BraveWalletService braveWalletService = mWalletModel.getBraveWalletService();
```

Comment:
If you want to use the mojom service outside domain model directly as a quick solution then can you use the reference coming from parent? (unless you gonna use AsyncInitializationActivity mentioned in the other comment).

Reply:
Yes, you're right!  I'll pick one of the two ways.